### ISSUE TITLE
Fix JSON encode/decode for demo frontend

### ISSUE THREAD

**Issue by nrutledge** [mlabs-haskell/plutus-use-cases#48]
The [lendex-frontend branch](https://github.com/mlabs-haskell/plutus-use-cases/tree/neil/lendex-frontend) contains a WIP demo frontend.

Currently the contract definitions are being fetched when the app is launched, however, the JSON encoding is failing with the following error in the console:

```
Uncaught Error: GET request to http://localhost:8081/http://localhost:8080/api/new/contract/definitions failed: : (Named "Array" (AtIndex 1 (AtKey "csrSchemas" (Named "Array" (AtIndex 0 (AtKey "argument" (Named "FormSchemaUnsupported" (AtKey "tag" (UnexpectedValue "FormSchemaObject")))))))))
    at Object.exports.error (foreign.js:8)
    at Object.$$throw [as throw] (index.js:19)
    at index.js:16
    at index.js:30
    at index.js:64
    at _run (foreign.js:278)
    at foreign.js:348
    at drain (foreign.js:120)
    at Object.enqueue (foreign.js:141)
    at foreign.js:339
```

The relevant types are:
- ContractSignatureResponse
- FunctionSchema
- EndpointDescription
- FormSchema

Below is a JSON response sample from http://localhost:8080/api/new/contract/definitions :
```
{
  "csrDefinition": "User",
  "csrSchemas": [
    {
      "endpointDescription": {
        "getEndpointDescription": "borrow"
      },
      "argument": {
        "tag": "FormSchemaObject",
        "contents": [
          [
            "borrow'amount",
            {
              "tag": "FormSchemaInteger"
            }
          ]
        ]
      }
    }
  ]
}
```

The FormSchema type below corresponds to the "argument" property of the JSON and was copied from the **plutus** repo (playground-common/src/Schema.hs):
```
data FormSchema
  = FormSchemaUnit
  | FormSchemaBool
  | FormSchemaInt
  | FormSchemaInteger
  | FormSchemaString
  | FormSchemaHex
  -- ^ A string that may only contain @0-9a-fA-F@
  | FormSchemaArray FormSchema
  | FormSchemaMaybe FormSchema
  | FormSchemaRadio (Array String)
  -- ^ A radio button with a list of labels.
  | FormSchemaTuple FormSchema FormSchema
  | FormSchemaObject (Array (Tuple String FormSchema))
  -- Blessed types that get their own special UI widget.
  | FormSchemaValue
  | FormSchemaSlotRange
  -- Exceptions.
  | FormSchemaUnsupported String
```

As you can see, the FormSchema doesn't exactly match the JSON format (specifically the "tag" property in the JSON. For this reason the **genericEncodeJson** / **genericDecodeJson** strategy currently being used is failing and will need to be modified so that the appropriate constructor is used based on the string contained in the "tag" field.

**Comment by Benjmhart** [comment]
this is fixed by https://github.com/mlabs-haskell/plutus-use-cases/pull/50

However the conditions of the fix have us holding some technical debt - for some reason it points to an issue with the `Eq` instance for `String`, this is too-depended-on to fail and in fact is itself a reference back to ffi over js `===` , which was exactly our fix.   rather puzzling.

the maintainer of `aeson-generic` requested that we drop in a minimal repro (which should work with try-purescript.org)

**Comment by nrutledge** [comment]
Additional fix has been applied for this issue: 09dd7649ba816545f3afdfc218fdc22ac7ff3b9e

It turns out the previous fix was causing all FormSchema properties to incorrectly decode to FormSchemaUnit. I've reverted the inclusion of the forked argonaut-generic.

In addition, I've  changed the `unwrapSingleArguments` encoding option from `false` to `true`. At this point the decoding appears to be behaving as expected.
